IPython/core/displayhook.py

```diff
@@ -187,7 +187,12 @@ def write_format_data(self, format_dict, md_dict=None):
                 # But avoid extraneous empty lines.
                 result_repr = '\n' + result_repr
 
-        print(result_repr)
+        try:
+            print(result_repr)
+        except UnicodeEncodeError:
+            # If a character is not supported by the terminal encoding replace
+            # it with its \u or \x representation
+            print(result_repr.encode(sys.stdout.encoding,'backslashreplace').decode(sys.stdout.encoding))
```

Comment:
Would it make sens to store in an private attribute wether you   print as is or `backslashreplace`, and directly use `backslashreplace` when the attribute is true ? 

I'm concern about printing things twice or at least partially twice.

Reply:
I can imagine your concerns. In cpython it seems to work well, but I didn't test it in other python implementations or with very large outputs. It is however the safest and fastest way I could think of.

A private attribute doesn't really make sense, you shouldn't need to set a private attribute to prevent ipython from crashing in my opinion.

Alternatives I could think of:
- always use 
```
print(result_repr.encode(sys.stdout.encoding,'backslashreplace').decode(sys.stdout.encoding))
```
This works and gives exactly the same output, but is considerably slower (30% slower in a quick test)

- Use print directly with UTF-8 encoding, otherwise encode and decode
```
if sys.stdout.encoding == 'UTF-8':
    print(a)
else:
    print(result_repr.encode(sys.stdout.encoding,'backslashreplace').decode(sys.stdout.encoding))
```
This is not measurably slower when the stdout encoding is utf-8, which is in most cases, it is slower on non utf-8 terminals when printing normal characters, but a bit faster when printing non-encodable characters.

I think the second solution is quite acceptable, no double print function call, still fast on utf-8 terminals, and a bit slower on other terminals. That shouldn't be a real issue, if the majority of your waiting time is in printing you might be printing too much.

What do you think?

Reply:
The default displayhook implementation uses a similar approach with try-except, see https://docs.python.org/3/library/sys.html#sys.displayhook

That strengthens me in the thought that my original approach is not so bad, and that partly printing before the UnicodeEncodeError is not a big risk.

Reply:
I was more thinking: 

```
if _always_replace:
    print(...encode(backslashreplace))
else:
   try:
      print():
    except:
      _always_replace=True
      print(replace)
```
So that basically once there is an issue it switch and stay in place; 

But Lets try that and we'll update if necessary.

Reply:
Ah, I see. That makes sense.
I guess anyway this issue will become less and less problematic as more and more systems should be updated to utf-8 encoding. 
Thanks for merging!